### PR TITLE
New lowering pass: loop rotation

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR adds a new lowering pass: loop rotation.

Loop rotation is a loop transformation that transforms the code like below
```python
for i in range(n):
  statement1(i)
  statement2(i)
  statement3(i)
  statement4(i)
```
into something like:
```python
if 0 < n:
  for i = 0:
    statement1(i)
    statement2(i)
for i in range(n):
  statement3(i)
  statement4(i)
  if i + 1 < n:
    statement1(i+1)
    statement2(i+1)
```
Right now, because existing predicates should already cover all illegal access, I am not materializing the `0 < n` and `i + 1 < n` predicates, so I am actually generating:
```python
for i = 0:
  statement1(i)
  statement2(i)
for i in range(n):
  statement3(i)
  statement4(i)
  if True:
    statement1(i+1)
    statement2(i+1)
```

Loop rotation is conceptually very simple regarding index and predicate lowering, and their changes are mostly mechanical. For the trivial loop `for i = 0`, our index and predicate lowering should already handle it correctly. For the exprs inside the `if i + 1 < n` block, the only change needed is to replace all the `loop->index()` with `loop->index() + 1`, which requires me to detect the `if i + 1 < n` block and mark its body as "rotated" during index and predicate lowering.

In order to use loop rotation, users can do `fusion.rotateLoop(tv, dim, {tv1, tv2})`, then during lowering, the loop `tv->axis(dim)` will be rotated, and the expressions allocating and computing `tv1` and `tv2` will be the rotated exprs.

### CODE REVIEW

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/lower_loop_rotation.h:8]: This note is a good starting point for review

File: third_party/nvfuser/test/test_gpu_loop_rotation.cpp
Comment (zasdfgbnm) [inline third_party/nvfuser/test/test_gpu_loop_rotation.cpp:204]: This is a test for https://github.com/csarofeen/pytorch/pull/2504

File: third_party/nvfuser/csrc/fusion.h
Comment (naoyam) [inline third_party/nvfuser/csrc/fusion.h:242]: Please comment on each of the parameters. Why is the last parameter a set of Statement rather than TensorView?

File: third_party/nvfuser/csrc/index_compute.h
Comment (naoyam) [inline third_party/nvfuser/csrc/index_compute.h]: nit: rotated_loops

File: third_party/nvfuser/csrc/kernel_ir.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/kernel_ir.cpp:602]: Why trivial when def is not null?

File: third_party/nvfuser/csrc/fusion.h
Comment (naoyam) [inline third_party/nvfuser/csrc/fusion.h:248]: Question on interface. This seems inconsistent compared to the rest of the API. For example, double buffering, which is kind of a similar concept, is done as: `tv->doubleBuffer();`. For loop rotation, why not `loop_tv->rotate(axis, selection)`?

File: third_party/nvfuser/csrc/type.h
Comment (naoyam) [inline third_party/nvfuser/csrc/type.h:56]: Please add a comment on `LoopRotation`. See above for the comments on the existing types.

**Review by naoyam — COMMENTED** [review]
Still working on, but some initial comments

File: third_party/nvfuser/csrc/predicate_compute.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/predicate_compute.cpp:626]: Can you please add a comment on why the last loop is a rotated loop when the predicate of an `IfThenElse` is `LoopRotation`?

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.h:108]: This predicate may be omitted when it's considered "simple". Is it safe to omit `if (0 < id1.extent()) {`?

**Comment by naoyam** [comment]
I get this build error (clang 13):

```
/home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/ir_cloner.h:67:10: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
        [this](auto&... x) { return std::make_tuple<Ts...>(clone(x)...); },
         ^
/home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/ir_cloner.h:49:22: note: in instantiation of function template specialization 'nvfuser::IrCloner::clone<nvfuser::TensorView *, long, std::unordered_set<nvfuser::Statement *>>' requested here
      copy.push_back(clone(p));
                     ^
/home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/fusion.cpp:83:40: note: in instantiation of function template specialization 'nvfuser::IrCloner::clone<std::tuple<nvfuser::TensorView *, long, std::unordered_set<nvfuser::Statement *>>>' requested here
  to->loop_rotation_param_ = ir_cloner.clone(from->loop_rotation_param_);
                                       ^
1 error generated.
```

File: third_party/nvfuser/csrc/index_compute.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/index_compute.cpp]: I wonder if the code could be simpler if we created a set of rotated loops as part of the `GpuLower` states instead of passing around `rotated_loop`. There seem a lot of places get changed just to pass `rotated_loop`, which could be avoided if we created it as part of the `GpuLower` states.

File: third_party/nvfuser/csrc/lower_index_compute.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_index_compute.cpp]: What does the code look like with both double buffering and loop rotation?

File: third_party/nvfuser/csrc/lower_index_compute.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_index_compute.cpp]: Do we need to create a new expr? Can we just check the lhs and rhs equality?

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.cpp]: nit: do you need this in addition to the below `handle` for `Expr*`?

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.cpp]: nit: `const std::unordered_set<Statement*>&`

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.cpp]: nit: Probably would be helpful to have some more context rather than just "nothing to do". "[Loop rotation] nothing to do"?

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.cpp:332]: Before this, each `expr` is in the body of the `fl` loop. Does this registration here also remove the expression from the `fl` loop body? Otherwise, doesn't it mean it exists in both before `fl` and its body?

Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp:332]: Ah, you're replacing `fl` with a new loop, so this shouldn't cause any problem? Maybe some comment would be helpful to understand.

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.cpp:341]: Is this the only place where a `LoopRotation` predicate is created?

File: third_party/nvfuser/csrc/fusion.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion.h:242]: Comment added. I use `Statement` to keep consistent with `lower_loop_rotation.h`. I want to reserve the right to specify expr, not just tv, although I am not using it.

File: third_party/nvfuser/csrc/index_compute.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/index_compute.h]: renamed

File: third_party/nvfuser/csrc/kernel_ir.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/kernel_ir.cpp:602]: Because
```C++
for (int i34 + 1 = 0; i34 + 1 < n, (i34 + 1)++)
```
does not make sense?

File: third_party/nvfuser/csrc/fusion.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion.h:248]: For double buffer, the scope is just limited to this tensor "I want to double buffer this tensor", but for loop rotation, the scope is an entire loop with all exprs inside the loop. So I belive it would be better to put the interface to more global places like fusion. Also, putting it to fusion is easier to access during lowering, because it is already there, and we don't need to write a pass to collect that information.

File: third_party/nvfuser/csrc/lower_index_compute.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_index_compute.cpp]: We are already creating lots of garbage exprs in expr simplifier, so I will not worry about memory saving here. Instead, I will just use whatever has the shortest code. In the future, we definitely need some garbage collection mechanism.

File: third_party/nvfuser/csrc/lower_index_compute.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_index_compute.cpp]: You can see `FusionLoopRotationDoubleBuffered_CUDA`, `FusionLoopRotationSelectDoubleBufferLoad_CUDA`, `FusionLoopRotationMultipleDoubleBuffer_CUDA`.

File: third_party/nvfuser/csrc/fusion.h
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion.h:248]: Hmm, I don't agree. It's a property on a loop, and a loop is represented by `IterDomain`. So, something like `loop_tv->axis(axis)->rotate(selection)` should also be enough. I don't understand why it needs to be a global property.

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: I don't think so, we need support for zero sized tensor.

File: third_party/nvfuser/csrc/index_compute.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/index_compute.cpp]: `rotated_loop` depend on the expr you are visiting. For example, if you have
```
statement1(0)
statement2(0)
for i ...
  statement3(i)
  statement4(i)
  if LoopRotation:
    statement1(i+1)
    statement2(i+1)
```
Then `for i...` is a rotated loop only if you are visiting the `statement1` and `statement2` inside the `if LoopRotation`, otherwise, for example if you are visiting `statement3` or `statement4`, then this `rotated_loop` will be empty

File: third_party/nvfuser/csrc/type.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/type.h:56]: Added

File: third_party/nvfuser/csrc/predicate_compute.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/predicate_compute.cpp:626]: added

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp]: fixed

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp]: added

File: third_party/nvfuser/csrc/lower_index_compute.cpp
Comment (naoyam) [reply third_party/nvfuser/csrc/lower_index_compute.cpp]: Then can we do just:

```
is_same = (rotated_loop.count(db_loop) && 
      cur_index->sameAs(SimplifyingIrBuilder::addExpr(db_loop->index(), db_loop->step())) || (cur_index == db_loop->index());
```

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp:341]: Yes

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp:332]: Your understanding is correct. Comment added.

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp]: No, removed. Thanks for catching!

File: third_party/nvfuser/csrc/index_compute.cpp
Comment (naoyam) [reply third_party/nvfuser/csrc/index_compute.cpp]: OK, makes sense.

File: third_party/nvfuser/csrc/lower_index_compute.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_index_compute.cpp]: You are absolutely right. I don't know what I was thinking when writing this code. Fixed.

File: third_party/nvfuser/csrc/fusion.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion.h:248]: Will this cause unterminated recursion in ir cloner?
```
loop_tv->id->selection->loop_tv->id->.....
```

**Comment by zasdfgbnm** [comment]
> I get this build error (clang 13):
> 
> ```
> /home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/ir_cloner.h:67:10: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
>         [this](auto&... x) { return std::make_tuple<Ts...>(clone(x)...); },
>          ^
> /home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/ir_cloner.h:49:22: note: in instantiation of function template specialization 'nvfuser::IrCloner::clone<nvfuser::TensorView *, long, std::unordered_set<nvfuser::Statement *>>' requested here
>       copy.push_back(clone(p));
>                      ^
> /home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/fusion.cpp:83:40: note: in instantiation of function template specialization 'nvfuser::IrCloner::clone<std::tuple<nvfuser::TensorView *, long, std::unordered_set<nvfuser::Statement *>>>' requested here
>   to->loop_rotation_param_ = ir_cloner.clone(from->loop_rotation_param_);
>                                        ^
> 1 error generated.
> ```

I think this is a bug of clang, but I changed the code to `this->clone` instead of `clone`. Hope this will make clang happy.

File: third_party/nvfuser/csrc/fusion.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion.h:248]: Looks like generally we should not make a tv directly or indirectly contain a pointer to tv, this could potentially cause unterminated recursion. `computeWith` is an exception, because we know that `computeWith` contains only consumer, which is guaranteed to never have a cyclic containing relationship.

**Comment by naoyam** [comment]
> > I get this build error (clang 13):
> > ```
> > /home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/ir_cloner.h:67:10: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
> >         [this](auto&... x) { return std::make_tuple<Ts...>(clone(x)...); },
> >          ^
> > /home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/ir_cloner.h:49:22: note: in instantiation of function template specialization 'nvfuser::IrCloner::clone<nvfuser::TensorView *, long, std::unordered_set<nvfuser::Statement *>>' requested here
> >       copy.push_back(clone(p));
> >                      ^
> > /home/nmaruyama/pytorch/debug2/third_party/nvfuser/csrc/fusion.cpp:83:40: note: in instantiation of function template specialization 'nvfuser::IrCloner::clone<std::tuple<nvfuser::TensorView *, long, std::unordered_set<nvfuser::Statement *>>>' requested here
> >   to->loop_rotation_param_ = ir_cloner.clone(from->loop_rotation_param_);
> >                                        ^
> > 1 error generated.
> > ```
> 
> I think this is a bug of clang, but I changed the code to `this->clone` instead of `clone`. Hope this will make clang happy.

Yeah, looks like so. I don't see any error with the change.

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp:341]: I think it's helpful to note that as part of the comment on `LoopRotation`

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: So, shouldn't we only omit the outer predicate (`if (0 < id1.extent())`) only when each expression is fully predicated with no omission? That sounds like messy logic, though...

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: Don't we already predicate all global and shared memory access? If it is access to register, I don't think this is a problem. We can do `T1[0] = sin(T0[0])` even if `T0` does not contain any value.

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp:341]: I have this in `type.h`:
```C++
// LoopRotation - Predicate added by loop rotation, currently always true.
```

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: The test in `test_gpu_loop_rotation.cpp` should ensure that they are predicated correctly

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: And I do have a note in the test about something that expr simplifier should change:
https://github.com/csarofeen/pytorch/pull/2500/files#diff-1b518fe84a6b24ba92a3dd37a4164a3f15fc6dc8f885094c155426f36cf3e615R105-R110

```C++
  // TODO: the predicate 0 < T0.size[0] in
  //  T1[i21]
  //     = T0[(T0.stride[1] * i29)];
  // is optimized to `true` by expr simplifier, due to
  // https://github.com/csarofeen/pytorch/blob/167718b6d06558395f86b6d25a68352168b86da2/third_party/nvfuser/csrc/expr_simplifier.cpp#L1115-L1126
  // This doesn't look very safe.
```

File: third_party/nvfuser/csrc/fusion.h
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion.h:248]: Ah, that's a good point.

I'm OK with the interface, although it still looks inconsistent to me and makes me a little uneasy as properties about a `Statement` may be defined in its container as well. Things like the IR printer as well as `sameAs` would need to look at the container as well, although in this particular case `sameAs` shouldn't care if rotated or not.

IMO, the design that makes the most sense is to have each TensorView to have a new attribute saying if one of its leaf IDs should be rotated. If multiple tensors are rotated on the same loop, then we just need to set the attribute of each of the rotated tensors. Since their axes are loop mapped, it indirectly indicates they are rotated together.

File: third_party/nvfuser/test/test_gpu_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/test/test_gpu_loop_rotation.cpp]: This looks inconsistent to me. Both are a fusion-wide operation, and one implicitly uses the fusion from a `FusionGuard`, but another is a method of `Fusion`. I'd say just `rotateLoop(tv4, -1, {tv1, tv2});` would align well with the existing APIs.

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: > Don't we already predicate all global and shared memory access? If it is access to register, I don't think this is a problem. We can do `T1[0] = sin(T0[0])` even if `T0` does not contain any value.

I don't think so. See https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/index_compute.cpp#L2657.

It only happens in a very trivial case with no split or merge.

**Comment by naoyam** [comment]
Just curious, can loop rotation be nested? Not that I think it's important.

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (naoyam) [inline third_party/nvfuser/csrc/lower_loop_rotation.cpp:180]: So, does this mean that even if a proper set of tensors are selected as rotated tensors when scheduling the fusion, the rotation could fail depending on how the expressions are ordered, and the failure doesn't cause the lowering to fail. 

Should this be the default behavior? Wouldn't you want the rotation happens whenever it's requested, and the lowering should fail if not? Assuming this optimization is only used for matmuls, shouldn't we prefer enforcing the rotation rather than gracefully ignoring the failure and silently causing slowdowns?

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: Yeah, I know that function. And my argument is, as long as we make sure that there is no out-of-bound memory access for global and shared memory, I don't really care if `if (0 < extent)` or `if (i + 1 < extent)` is checked or not. **So the code with and without this predicate will not be equivalent, but I don't care if it is equivalent or not.** Do we consider this claim safe? I started this PR with all these predicates, but I eventually removed them because I prefer to read the code that without it (this simplification is mostly for readability, not for perf, I didn't check perf of it).

File: third_party/nvfuser/csrc/fusion.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion.h:248]: I don't consider the infomation "XXX loop need to rotate XXX exprs" as an attribute of a tensor or an ID. I more consider it as a lowering hint/requirement of the kernel. And I don't want the selection to be limited to tensors. I want to be more flexible, and using that same interface, we can also select exprs, instead of tensors.

And because fusion schedulers don't have access to kernel structure, I generally don't think it makes sense to let fusion schedulers to provide loop rotation information. For example, currently, if I select the double buffer load for rotation, it will not only rotate the double buffer main loop, but it will also rotate the double buffer prologue loop. And the current interface does not allow control of that. The only reason I designed this interface is to make it convenient for writing unit tests. I don't expect this interface to be a utility like `computeAt` which will be used widely in schedulers. I consider this interface as a lightweight hacking point for unit tests and matmul scheduler to talk to lowering, so I don't want it to be buried deep in our design. And I do think we need to revisit the design of interface when we get another user that is not a matmul or unit test.

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: I agree with this:

> as long as we make sure that there is no out-of-bound memory access for global and shared memory, I don't really care if if (0 < extent) or if (i + 1 < extent) is checked or not.

What I'm asking is in some cases the predicate of an expr is omitted based on the assumption that the expr is located inside a loop. Since this transformation moves an expression out of a loop, this assumption may not be held, so my question is are we making sure there is no out-of-bound memory access for global and shared memory?

File: third_party/nvfuser/csrc/lower_loop_rotation.cpp
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.cpp:180]: Yeah, this makes sense. Changed.

**Comment by zasdfgbnm** [comment]
> Just curious, can loop rotation be nested? Not that I think it's important.

I don't think so. To support nested loop rotation, we need to carefully handle the `registerReplace` and `registerInsertBefore` to make sure it does not change the abandoned main loop.

File: third_party/nvfuser/csrc/fusion.h
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion.h:248]: Why tensors are not enough? Why do you want to be able to specify exprs? I'm thinking about what the interface should be for a fusion. 

For the case like double buffering and loop rotation, I see that you may want to select a particular expr among all exprs defining the same tensor, i.e., the expr appearing in the main loop but not in the prologue. But that's not a `Fusion` anymore. It seems it only makes sense in the lowering phase.

**Comment by naoyam** [comment]
> > Just curious, can loop rotation be nested? Not that I think it's important.
> 
> I don't think so. To support nested loop rotation, we need to carefully handle the `registerReplace` and `registerInsertBefore` to make sure it does not change the abandoned main loop.

OK. Maybe add a comment about it?

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: Are you saying that `the predicate of an expr is omitted based on the assumption that the expr is located inside a loop`, and `canOmitStopPredicate` is the code that decide whether it should be omitted? I don't see any problem here: `canOmitStopPredicate` only omit the predicate if `stop_index + stop_offset < extent` is proved to be true, and if this is proved to be true, `0 < extent` or `i + 1 < extent` will automatically be satisfied?

File: third_party/nvfuser/csrc/fusion.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion.h:248]: I am not strong on putting it as an attribute of fusion. But it does have to be something that is preserved during fusion copy, so it can not be a compilation parameter. And I don't want to do interface designing in this PR, because I think it makes more sense to do this work when there are more use cases. For this PR, I want this interface to be something that is simple and easy to remove, so that, in the future, it can be easy to refactored. That said, the approach with least design, i.e. directly putting the argument of `rotateLoops` without any change as an attribute of fusion is the simplest approach, so I prefer this approach.

In this PR, it does not has to be able to specify expr. But long term, if the argument of `rotateLoops` is no longer passed from fusion, but somewhere that makes more sense, we will be able to provide the `Expr*` from the previous pass. So it does make sense to let the argument of `rotateLoops` to be able to specify expr. And I want to keep the interface simple, and take the designless approach, which means in the fusion, we should also be able to specify expr.

**Comment by zasdfgbnm** [comment]
> > > Just curious, can loop rotation be nested? Not that I think it's important.
> > 
> > 
> > I don't think so. To support nested loop rotation, we need to carefully handle the `registerReplace` and `registerInsertBefore` to make sure it does not change the abandoned main loop.
> 
> OK. Maybe add a comment about it?

Actually, it is pretty easy to support it with very little extra work: I can just rotate one loop each pass.

I updated this PR to support this in https://github.com/csarofeen/pytorch/pull/2500/commits/67f7df347d86fdf035323046aa083c2c7635b6b5

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: The omission is done based on the assumption that:

```
stop_index >= 0
stop_index < extent
```

This assumption is valid when `index_simple` is true (https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/index_compute.cpp#L2661). The analysis is that when the index has no defining expression, it must be true that the stop index should be just a loop index, so it is guaranteed to range from 0 to extent.

If this expr is placed outside the loop, we can no longer have the assumption. More specifically, when `extent` is 0, the loop should never execute, so the omission of the predicate of the expression shouldn't matter. However, when the expression is moved out of the loop, there should be some predicate not to execute the expression when `extent` is 0.

Am I missing something? I actually am not familiar with how extent-zero domains are handled, but it seems to me omitting the predicate is wrong in that case, granted it's super uncommon to have extent-zero domains and simple indexing.

File: third_party/nvfuser/csrc/fusion.h
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion.h:248]: OK, fair enough. The only remaining comment regarding this is https://github.com/csarofeen/pytorch/pull/2500#discussion_r1119490970

File: third_party/nvfuser/csrc/lower_loop_rotation.h
Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: I see what you mean. And I think what you said makes sense. Let me try to play with `canOmitStopPredicate` to see if we can avoid this issue.

> I actually am not familiar with how extent-zero domains are handled

I don't think extent-zero domains are handled. They are just treated no differently from other extents, so there might be bugs somewhere in our system.

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: Do you think this commit fixes the problem? https://github.com/csarofeen/pytorch/pull/2500/commits/80d7763e141898771c64a5a9eea2d587dd509b62

Comment (naoyam) [reply third_party/nvfuser/csrc/lower_loop_rotation.h:108]: Yes, I think so. Thanks for the fix. Just left one minor comment.

**Review by naoyam — APPROVED** [review]
Approving now. Thanks for the updates. This is still not something I like, https://github.com/csarofeen/pytorch/pull/2500#discussion_r1119490970, but it's a question on code style, and it shouldn't block the PR.

**Pull request merged**
